File: src/VoiceBuddy.App/Services/Resampler.cs
namespace VoiceBuddy.Services;

/// <summary>
/// Linear resampler for mono float PCM. Good enough for speech in the 16 kHz Whisper pipeline;
/// a polyphase filter would be better but not worth the complexity before the STT stage exists.
/// </summary>
public static class Resampler
{
    public static float[] Linear(float[] input, int inputRate, int outputRate)
    {
        if (inputRate == outputRate || input.Length == 0) return input;
        double ratio = (double)inputRate / outputRate;
        int outLen = (int)(input.Length / ratio);
        if (outLen <= 0) return Array.Empty<float>();

        var output = new float[outLen];
        for (int i = 0; i < outLen; i++)
        {
            double srcIdx = i * ratio;
            int idx0 = (int)srcIdx;
            int idx1 = idx0 + 1 < input.Length ? idx0 + 1 : input.Length - 1;
            float t = (float)(srcIdx - idx0);
            output[i] = input[idx0] * (1 - t) + input[idx1] * t;
        }
        return output;
    }
}

File: src/VoiceBuddy.Chrome/lib/resampler.js
// Linear resampler for mono Float32 PCM — direct port of
// src/VoiceBuddy.App/Services/Resampler.cs. Good enough for speech.
export function linearResample(input, inputRate, outputRate) {
  if (inputRate === outputRate || input.length === 0) return input;
  const ratio = inputRate / outputRate;
  const outLen = Math.floor(input.length / ratio);
  if (outLen <= 0) return new Float32Array(0);

  const out = new Float32Array(outLen);
  for (let i = 0; i < outLen; i++) {
    const srcIdx = i * ratio;
    const idx0 = srcIdx | 0;
    const idx1 = idx0 + 1 < input.length ? idx0 + 1 : input.length - 1;
    const t = srcIdx - idx0;
    out[i] = input[idx0] * (1 - t) + input[idx1] * t;
  }
  return out;
}
